README.md

```diff
@@ -1,4 +1,151 @@
+$ ./run-command.sh 8983 start
+$ ./run-command.sh 8983 stop
+```
+
+By default, the profiling frequency is 100Hz (every 10ms). Here is a sample of
+the output printed to the Java application's terminal:
+
+```
+--- Execution profile ---
+Total:                   687
+Unknown (native):        1 (0.15%)
+
+Samples: 679 (98.84%)
+    [ 0] Primes.isPrime
+    [ 1] Primes.primesThread
+    [ 2] Primes.access$000
+    [ 3] Primes$1.run
+    [ 4] java.lang.Thread.run
+
+... a lot of output omitted for brevity ...
+
+         679 (98.84%) Primes.isPrime
+           4 (0.58%)  __do_softirq
+
+... more output omitted ...
+```
+
+This indicates that the hottest method was `Primes.isPrime`, and the hottest
+call stack leading to it comes from `Primes.primesThread`.
+
+## Generating Flame Graphs
+
+To generate flame graphs or other visualizations from the collected profiling
+information, you will need to dump the raw collected traces to a file and then
+post-process it. The following example uses Brendan Gregg's
+[FlameGraph](https://github.com/BrendanGregg/FlameGraph) scripts, but a similar
+solution can be tailored to other visualization tools.
+
+```
+$ jps
+9234 Jps
+8983 Computey
+$ ./run-command.sh 8983 interval:1,start
+$ ./run-command.sh dumpRawTraces:/tmp/traces.txt
+$ FlameGraph/stackcollapse-ljp.awk /tmp/traces.txt > /tmp/folded.stacks
+$ FlameGraph/flamegraph.pl /tmp/folded.stacks > /tmp/flamegraph.svg
+```
+
+## All Agent Options
+
+The following is a complete list of the command-line options accepted by the
+agent. Some options can be combined, e.g. the profiling interval can be
+provided at the same time as the `start` command, by separating them with
+commas: `interval:1,start`.
+
+* `interval:N` - sets the profiling interval, in milliseconds.
+
+* `frameBufferSize:N` - sets the frame buffer size, in the number of Java
+method ids that should fit in the buffer. If you receive messages about an
+insufficient frame buffer size, increase this value from the default.
+
+* `start` - starts profiling.
+
+* `stop` - stops profiling and prints the report.
+
+* `dumpRawTraces:filename` - stops profiling and dumps the raw trace information
+to the specified file.
+
+The format of the raw trace file is a collection of call stacks, where the first
+line is the number of appearances of that call stack, the number of frames in
+the call stack, followed by the call stack with each frame on a separate line.
+For example:
+
+```
+1056	5	Primes::isPrime
+		Primes::primesThread
+		Primes::access$000
+		Primes$1::run
+		java.lang.Thread::run
+```
+
+This stack appeared 1056 times in the trace. This is the same stack trace format
+used by the [Lightweight Java Profiler](https://code.google.com/archive/p/lightweight-java-profiler/),
+and can be easily used for generating flame graphs.
+
+## Restrictions/Limitations
+
+* Accessing `perf_events` requires root privileges. Because the agent is loaded
+into the Java process, **the Java process must run as root**.
```

Comment:
Not really. PERF_COUNT_SW_CPU_CLOCK counter (used by default) does not require root privileges. async-profiler opens per-thread events that are also available for unprivileged users. Before Linux 4.6 it was even possible to get kernel callchains. Since 4.6 only user-mode profiling is available by default, but it is possible to override this by setting /proc/sys/kernel/perf_event_paranoid to 1 or below. Also /proc/sys/kernel/kptr_restrict should be 0 to allow resolving kernel symbols by non-root.
In fact, I could successfully get full-stack profiles under non-root account on all systems I've tested, e.g. CentOS 7 / 3.10 kernel or Ubuntu 16.04 / 4.4 kernel.